packages/upload-client/test/blob.test.js

```diff
@@ -177,6 +187,138 @@ describe('Blob.add', () => {
+    const agent = await Signer.generate()
+    const bytes = await randomBytes(128)
+
+    const proofs = [
+      await BlobCapabilities.add.delegate({
+        issuer: space,
+        audience: agent,
+        with: space.did(),
+        expiration: Infinity,
+      }),
+    ]
+
+    const service = mockService({
+      ucan: {
+        conclude: provide(UCAN.conclude, () => {
+          return { ok: { time: Date.now() } }
+        }),
+      },
+      blob: {
+        // @ts-ignore Argument of type
+        add: provide(BlobCapabilities.add, ({ invocation }) => {
+          return setupBlobAddSuccessResponse(
+            { issuer: space, audience: agent, with: space, proofs },
+            invocation
+          )
+        }),
+      },
+    })
+
+    const server = Server.create({
+      id: serviceSigner,
+      service,
+      codec: CAR.inbound,
+      validateAuthorization,
+    })
+    const connection = Client.connect({
+      id: serviceSigner,
+      codec: CAR.outbound,
+      channel: server,
+    })
+
+    await assert.rejects(
+      Blob.add(
+        { issuer: agent, with: space.did(), proofs, audience: serviceSigner },
+        bytes,
+        {
+          connection,
+          retries: 0,
+          fetch: async (url) => {
+            // @ts-ignore Parameter
+            if (!url.pathname) {
+              return await fetch(url)
+            }
+            throw new Server.Failure('boom')
+          },
+        }
+      ),
+      {
+        message: 'failed blob/add invocation',
+      }
+    )
+  })
+
+  it('throws when it there is no blob/accept receipt', async () => {
+    const space = await Signer.generate()
+    const agent = await Signer.generate()
+    const bytes = await randomBytes(128)
+
+    const proofs = [
+      await BlobCapabilities.add.delegate({
+        issuer: space,
+        audience: agent,
+        with: space.did(),
+        expiration: Infinity,
+      }),
+    ]
+
+    const service = mockService({
+      ucan: {
+        conclude: provide(UCAN.conclude, () => {
+          return { ok: { time: Date.now() } }
+        }),
+      },
+      blob: {
+        // @ts-ignore Argument of type
+        add: provide(BlobCapabilities.add, ({ invocation }) => {
+          return setupBlobAddSuccessResponse(
+            { issuer: space, audience: agent, with: space, proofs },
+            invocation
+          )
+        }),
+      },
+    })
+
+    const server = Server.create({
+      id: serviceSigner,
+      service,
+      codec: CAR.inbound,
+      validateAuthorization,
+    })
+    const connection = Client.connect({
+      id: serviceSigner,
+      codec: CAR.outbound,
+      channel: server,
+    })
+
+    await assert.rejects(
+      Blob.add(
+        { issuer: agent, with: space.did(), proofs, audience: serviceSigner },
+        bytes,
+        {
+          connection,
+          retries: 0,
+          fetch: async (url) => {
+            // @ts-ignore Parameter
+            if (!url.pathname) {
```

Comment:
There is always a `pathname` in URL no? It's `/` if no path...

Reply:
I'm guessing this works because you're only using this value for `fetch` in the receipts call.

If yes, the `fetch` option should not be documented in the types since it's not used for all requests made by `Blob.add()`.

This is convenient for testing but I think spinning up an actual server to do this would be better.

Reply:
this is gone now